### PR TITLE
zoom-us: fix web login

### LINKED ISSUE #75903: zoom-us: "Sign in with google" causes browser to open, without any link
**Describe the bug**
SSO login in zoom just opens the default browser that loads the default homepage

**To Reproduce**
1. Open zoom-us
2. Click sign-in
3. Click sign in with google

**Expected behavior**
A browser should open _pointed at the Google SSO page_

**Screenshots**

**Additional context**
Seems like the argument isn't passed properly to the browser. Possibly wrapper issue or the default-browser-command-thing is somehow swallowing the parameter.

**Metadata**
 - system: `"x86_64-linux"`
 - host os: `Linux 4.19.89, NixOS, 20.03pre-git (Markhor)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.1`
 - channels(root): `"nixos-20.03pre205780.d942688fc13"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: zoom-us
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by danbst**
Fixes https://github.com/NixOS/nixpkgs/issues/75903
Partially helps with https://github.com/NixOS/nixpkgs/issues/73532, because it is now possible to disable `embeddedBrowserForSSOLogin` and login via browser

cc @dtzWill @tadfisher @mkg20001 @pbogdan 

cc @worldofpeace @bjornfor @ttuegel maybe it is possible to reuse existing wrapper, I haven't found a way.

### CODE REVIEW

**Review by bjornfor — APPROVED** [review]
Looks good to me. But please prefix 2nd commit with something like "maintainers: ".

**Review by pbogdan — APPROVED** [review]
LGTM, confirmed "Sign in with Google" works on my end.

Would it be feasible at all to have this included in some form on 19.09 as well? "Sign in with Google" would I believe require bumping zoom to a newer version too though...

**Comment by danbst** [comment]
@pbogdan yeah, I'll backport it, perhaps together with Zoom update

File: pkgs/applications/networking/instant-messengers/zoom-us/default.nix [hunk 1]
```diff
@@ -81,7 +81,7 @@ in mkDerivation {
     # Desktop File
     cp ${desktopIntegration}/us.zoom.Zoom.desktop $out/share/applications
     substituteInPlace $out/share/applications/us.zoom.Zoom.desktop \
-        --replace "Exec=zoom" "Exec=$out/bin/zoom-us"
+        --replace "Exec=zoom" "Exec=$out/share/zoom-us/ZoomLauncher"
```
Comment (worldofpeace) [inline pkgs/applications/networking/instant-messengers/zoom-us/default.nix]: Why does this have to change if you still provide `/bin/zoom-us`?

File: pkgs/applications/networking/instant-messengers/zoom-us/default.nix [hunk 2]
```diff
@@ -95,9 +95,28 @@ in mkDerivation {
         cp $icon $out/share/icons/hicolor/$path/apps/us.zoom.Zoom.png
     done
 
-    ln -s $out/share/zoom-us/zoom $out/bin/zoom-us
+
+    # Do the manual Qt wrapping, because existing hook doesn't support our usecase
+    qtWrapperArgs+=(--prefix XDG_DATA_DIRS : "$out/share/zoom-us")
```
Comment (worldofpeace) [inline pkgs/applications/networking/instant-messengers/zoom-us/default.nix]: can use use the nix attr below for this, or does it not work? (I believe you can use placeholder)

Comment (worldofpeace) [reply pkgs/applications/networking/instant-messengers/zoom-us/default.nix]: After inspecting the produced wrappers they're missing an entry to XDG_DATA_DIRS for `$out/share` which **needs** to be in these wrappers too. I see code in wrapQtAppsHook for that, so this confuses me.

Can you also add a comment why we need the `zoom-us` subdir in XDG_DATA_DIRS?

Comment (worldofpeace) [reply pkgs/applications/networking/instant-messengers/zoom-us/default.nix]: I also believe you're wrapping in the wrong phase, can this be done in `preFixup`?

File: pkgs/applications/networking/instant-messengers/zoom-us/default.nix [hunk 3]
```diff
@@ -95,9 +95,28 @@ in mkDerivation {
...
+
+    find "$out/share/zoom-us" ! -type d -executable -print0 | while IFS= read -r -d "" file; do
+      patchelf --print-interpreter "$file" >/dev/null 2>&1 || continue
```
Comment (worldofpeace) [inline pkgs/applications/networking/instant-messengers/zoom-us/default.nix]: doesn't autoPatchelfHook make this unneeded?

File: pkgs/applications/networking/instant-messengers/zoom-us/default.nix [hunk 4]
```diff
@@ -95,9 +95,28 @@ in mkDerivation {
...
+
+    # it has to pass out var into wrapper
+    cat > $out/bin/zoom-us <<EOF
+    #!${bash}/bin/bash
+    (
+      cd $out/share/zoom-us
+      exec ./ZoomLauncher
+    )
+    EOF
+    chmod +x $out/bin/zoom-us
```
Comment (worldofpeace) [inline pkgs/applications/networking/instant-messengers/zoom-us/default.nix]: this is exactly what makeWrapper does. I don't see any reason why you shouldn't first cd to zoom directory with all the produced wrappers (that being `zoom` `zopen` and `ZoomLauncher`).

```
qtWrapperArgs = [
  ...
 ''--run "cd ${placeholder "out"}/share/zoom-us"''
];

Comment (worldofpeace) [reply pkgs/applications/networking/instant-messengers/zoom-us/default.nix]: comment also why that is needed.

**Review by worldofpeace — CHANGES_REQUESTED** [review]
This change is a bit confusing to me, I'm rather familiar with the wrapping setup hooks as well.

File: pkgs/applications/networking/instant-messengers/zoom-us/default.nix (hunk 1 above)
Comment (danbst) [reply pkgs/applications/networking/instant-messengers/zoom-us/default.nix]: actually, you're right here.

File: pkgs/applications/networking/instant-messengers/zoom-us/default.nix (hunk 3 above)
Comment (danbst) [reply pkgs/applications/networking/instant-messengers/zoom-us/default.nix]: right! thanks

**Comment by danbst** [comment]
@worldofpeace thanks a lot for review, didn't know about `placeholder` (wanted it to exist!).

There was a little problem with passing `--run "cd blabla"` into qt wrapper. It splits it into 3 args `--run`, `"cd` and `blabla"` which isn't what I wanted. I was not able to solve it in `wrap-qt-apps-hook.sh`.

And the other problem comes from fact that zoom binaries live in directory, which isn't included into qt wrapper `targetDirs`. Some "custom dirs" qt wrapper arg missing? Dunno

File: pkgs/applications/networking/instant-messengers/zoom-us/default.nix (hunk 2 above)
Comment (danbst) [reply pkgs/applications/networking/instant-messengers/zoom-us/default.nix]: @worldofpeace 
> I also believe you're wrapping in the wrong phase, can this be done in preFixup?

changed to `postFixup`, as I see there are some hooks added in `preFixup` in wrap-qt-setup-hook.sh. I don't know what will be the run order, but postFixup will definitely run after.

File: maintainers/scripts/update.nix [hunk 5]
```diff
@@ -105,7 +105,7 @@ let
     to run all update scripts for all packages that lists \`garbas\` as a maintainer
     and have \`updateScript\` defined, or:
 
-        % nix-shell maintainers/scripts/update.nix --argstr package garbas
+        % nix-shell maintainers/scripts/update.nix --argstr package gnome3
```
Comment (worldofpeace) [inline maintainers/scripts/update.nix]: I actually thing several things aren't correct here. This should be like `maintainer $maintainer_attr`, and to run for a specific package below should be `package`. This is out of scope and has various other inaccuracy, can you drop the commit?

File: pkgs/applications/networking/instant-messengers/zoom-us/default.nix [hunk 6]
```diff
@@ -94,19 +94,35 @@ in mkDerivation {
...
+    # ^^ unfortunately, breaks run arg into multiple array elements, due to
+    # some bad array propagation. We'll do that in bash below
   ];
 
+  postFixup = ''
+    # Zoom expects "zopen" executable (needed for web login) to be present in CWD. Or does it expect
+    # everybody runs Zoom only after cd to Zoom package directory? Anyway, :facepalm:
+    qtWrapperArgs+=( --run "cd ${builtins.placeholder "out"}/share/zoom-us" )
+
+    for app in ZoomLauncher zopen zoom; do
```
Comment (worldofpeace) [inline pkgs/applications/networking/instant-messengers/zoom-us/default.nix]: :facepalm also for zopen
```
❯ ./result/share/zoom-us/zopen --help
xdg-open - opens a file or URL in the user's preferred application
```

it's literally xdg-open.

File: pkgs/applications/networking/instant-messengers/zoom-us/default.nix [hunk 7]
```diff
@@ -94,19 +94,35 @@ in mkDerivation {
...
     ''--prefix LD_PRELOAD : ${libv4l}/lib/libv4l/v4l2convert.so''
+    # --run "cd ${builtins.placeholder "out"}/share/zoom-us"
+    # ^^ unfortunately, breaks run arg into multiple array elements, due to
+    # some bad array propagation. We'll do that in bash below
   ];
 
+  postFixup = ''
+    # Zoom expects "zopen" executable (needed for web login) to be present in CWD. Or does it expect
+    # everybody runs Zoom only after cd to Zoom package directory? Anyway, :facepalm:
+    qtWrapperArgs+=( --run "cd ${builtins.placeholder "out"}/share/zoom-us" )
```
Comment (worldofpeace) [inline pkgs/applications/networking/instant-messengers/zoom-us/default.nix]:
```suggestion
    qtWrapperArgs+=( --run "cd ${placeholder "out"}/share/zoom-us" )
```

I don't think you even need to use that attrpath in **nixpkgs** for that function.

**Comment by worldofpeace** [comment]
https://github.com/NixOS/nixpkgs/issues/75903#issuecomment-567503573 is reason enough for me to believe you should do this. Though this is the constant drawback with unfree software :rofl:, you're never going to know what exactly it's doing easily.

File: maintainers/scripts/update.nix (hunk 5 above)
Comment (danbst) [reply maintainers/scripts/update.nix]: no, it is real typo. Your suggested variant is already listed https://github.com/NixOS/nixpkgs/blob/master/maintainers/scripts/update.nix#L103-L111

File: pkgs/applications/networking/instant-messengers/zoom-us/default.nix (hunk 6 above)
Comment (danbst) [reply pkgs/applications/networking/instant-messengers/zoom-us/default.nix]: It uses cross-platform https://doc.qt.io/qt-5/qdesktopservices.html#openUrl, which uses `xdg-open` under the hood (confirmed with strace)

File: maintainers/scripts/update.nix (hunk 5 above)
Comment (danbst) [reply maintainers/scripts/update.nix]: but I found that my suggestion was bad as well. I've changed the suggestion and verified it runs

**Review by worldofpeace — APPROVED** [review]
Looks about correct :+1:

**Comment by danbst** [comment]
cherried to https://github.com/NixOS/nixpkgs/commit/33bacd39b946bb0c5b449a81779f953e466c12f4

Also, updated Zoom to latest in release-19.09
